### ISSUE TITLE
"webpacker:yarn_install" Task removes dev dependencies

### ISSUE THREAD

**Issue by odlp** [rails/webpacker#1330]
On CircleCI, with a Rails 4.2 app, our RSpec test suite precompiles assets. This triggers a `webpacker:yarn_install` and `webpacker:compile`:

https://github.com/rails/webpacker/blob/54c0a64adac57f05760cbd3fed5222d72e39593f/lib/tasks/webpacker/compile.rake#L11-L19

Unfortunately the `webpacker:yarn_install` task isn't environment-aware:

https://github.com/rails/webpacker/blob/54c0a64adac57f05760cbd3fed5222d72e39593f/lib/tasks/webpacker/yarn_install.rake#L4

The `--production` flag ends up removing the JS dev dependencies. On CI this means we'd have to re-run `yarn install` before running JS unit tests or redefine the Rake task.

Rough overview of the steps on CI:

```sh
yarn install
ls node_modules | wc -l # 966

bundle exec rspec
ls node_modules | wc -l # 666

yarn test

# yarn run v1.5.1
# $ RAILS_ENV=test node_modules/.bin/karma start --single-run
# sh: node_modules/.bin/karma: No such file or directory
```

**Comment by luccasmaso** [comment]
Nice one, spent some hours until discover this )=

**Comment by odlp** [comment]
As a temporary workaround one can redefine the Rake task:

```ruby
# lib/tasks/webpacker.rake

# Redefining task so we don't remove the dev dependencies when running on CI
Rake::Task["webpacker:yarn_install"].clear

namespace :webpacker do
  desc "Support for older Rails versions. Install all JavaScript dependencies as specified via Yarn"
  task :yarn_install do
    if ENV["RAILS_ENV"] == "production"
      system "yarn install --no-progress --frozen-lockfile --production"
    else
      system "yarn install --no-progress"
    end
  end
end
```

**Comment by odlp** [comment]
Closing - fixed in #1331
